SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ClosureSpecialization.swift

```diff
@@ -0,0 +1,716 @@
+/// patterns of Swift Autodiff.
+///
+/// The compiler performs reverse-mode differentiation on functions marked with `@differentiable(reverse)`. In doing so,
+/// it generates corresponding VJP and Pullback functions, which perform the forward and reverse pass respectively. You
+/// can think of VJPs as functions that "differentiate" an original function and Pullbacks as the calculated
+/// "derivative" of the original function. 
+/// 
+/// VJPs always return a tuple of 2 values -- the original result and the Pullback. Pullbacks are essentially a chain 
+/// of closures, where the closure-contexts are implicitly used as the so-called "tape" during the reverse
+/// differentiation process. It is this chain of closures contained within the Pullbacks that this optimization aims
+/// to optimize via closure specialization.
+///
+/// The code patterns that this optimization targets, look similar to the one below:
+/// ``` swift
+/// 
+/// // Since `foo` is marked with the `differentiable(reverse)` attribute the compiler
+/// // will generate corresponding VJP and Pullback functions in SIL. Let's assume that
+/// // these functions are called `vjp_foo` and `pb_foo` respectively.
+/// @differentiable(reverse) 
+/// func foo(_ x: Float) -> Float { 
+///   return sin(x)
+/// }
+///
+/// //============== Before optimization ==============// 
+/// // VJP of `foo`. Returns the original result and the Pullback of `foo`.
+/// sil @vjp_foo: $(Float) -> (originalResult: Float, pullback: (Float) -> Float) { 
+/// bb0(%0: $Float): 
+///   // vjp_sin returns the original result `sin(x)`, and the pullback of sin, `pb_sin`.
+///   // `pb_sin` is a closure.
+///   %(originalResult, pb_sin) = apply @vjp_sin(%0): $(Float) -> (Float, (Float) -> Float)
+///  
+///   %pb_foo = function_ref @pb_foo: $@convention(thin) (Float, (Float) -> Float) -> Float
+///   %partially_applied_pb_foo = partial_apply %pb_foo(%pb_sin): $(Float, (Float) -> Float) -> Float
+///  
+///   return (%originalResult, %partially_applied_pb_foo)
+/// }
+///
+/// // Pullback of `foo`. 
+/// //
+/// // It receives what are called as intermediate closures that represent
+/// // the calculations that the Pullback needs to perform to calculate a function's
+/// // derivative.
+/// //
+/// // The intermediate closures may themselves contain intermediate closures and
+/// // that is why the Pullback for a function differentiated at the "top" level
+/// // may end up being a "chain" of closures.
+/// sil @pb_foo: $(Float, (Float) -> Float) -> Float { 
+/// bb0(%0: $Float, %pb_sin: $(Float) -> Float): 
+///   %derivative_of_sin = apply %pb_sin(%0): $(Float) -> Float 
+///   return %derivative_of_sin: Float
+/// }
+///
+/// //============== After optimization ==============// 
+/// sil @vjp_foo: $(Float) -> (originalResult: Float, pullback: (Float) -> Float) { 
+/// bb0(%0: $Float): 
+///   %1 = apply @sin(%0): $(Float) -> Float 
```

Comment:
It is a bit misleading here. In order to do this you'd need to inline `vjp_sin`, why `specialize` then?. How else you'd know the pullback given the vjp?

Reply:
> It is a bit misleading here. 
Perhaps I can mention that vjp_sin must have been inlined outside of the optimization, to make things clearer.

Reply:
Plus you can just show the body of `vjp_sin`.